!wget https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/refs/heads/master/cohorts/2024/01-intro/homework.md

## Homework

### Set up the environment

You need to install Python, NumPy, Pandas, Matplotlib and Seaborn. For that, you can use the instructions from
[06-environment.md](../../../01-intro/06-environment.md).

### Q1. Pandas version

What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

In [14]:
import pandas as pd
f'pandas version installed : { pd.__version__}'


'pandas version installed : 2.2.2'

### Getting the data 

For this homework, we'll use the Laptops Price dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv).


!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv

### Q2. Records count

How many records are in the dataset?

In [13]:
df = pd.read_csv('/workspaces/ml-zoomcamp/01-intro/laptops.csv')

df.shape

f"number of records : {len(df)}"

'number of records : 2160'

### Q3. Laptop brands

How many laptop brands are presented in the dataset?

- 12
- 27
- 28
- 2160

In [27]:
f'count of laptop brands: {len(df['Brand'].unique())}'

'count of laptop brands: 27'

### Q4. Missing values

How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3


In [29]:
df.isnull().sum()

# 2 columns have missing values

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

### Q5. Maximum final price

What's the maximum final price of Dell notebooks in the dataset?

- 869
- 3691
- 3849
- 3936

In [38]:
max_value_dell = df[df['Brand'] == 'Dell']['Final Price'].max()
print(f"Maximum value for Dell: {max_value_dell}")


Maximum value for Dell: 3936.0


### Q6. Median value of Screen

1. Find the median value of `Screen` column in the dataset.
2. Next, calculate the most frequent value of the same `Screen` column.
3. Use `fillna` method to fill the missing values in `Screen` column with the most frequent value from the previous step.
4. Now, calculate the median value of `Screen` once again.


In [42]:
import pandas as pd

# 1. Find the median value of `Screen` column before filling NaN values
initial_median = df['Screen'].median()
print("Initial median value of Screen:", initial_median)

# 2. Find the most frequent value (mode) of `Screen`
most_frequent_value = df['Screen'].mode()[0]
print("Most frequent value of Screen:", most_frequent_value)

# 3. Fill missing values in `Screen` with the most frequent value
df['Screen'] = df['Screen'].fillna(most_frequent_value)

# 4. Recalculate the median value of `Screen` after filling NaN values
new_median = df['Screen'].median()
print("New median value of Screen after filling missing values:", new_median)


Initial median value of Screen: 15.6
Most frequent value of Screen: 15.6
New median value of Screen after filling missing values: 15.6


### Q7. Sum of weights

1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns `RAM`, `Storage`, `Screen`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the sum of all the elements of the result?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.


- 0.43
- 45.29
- 45.58
- 91.30


In [43]:
import pandas as pd
import numpy as np


# 1. Select all the "Innjoo" laptops
innjoo_laptops = df[df['Brand'] == 'Innjoo']

# 2. Select only columns `RAM`, `Storage`, `Screen`
selected_columns = innjoo_laptops[['RAM', 'Storage', 'Screen']]

# 3. Get the underlying NumPy array (X)
X = selected_columns.to_numpy()

# 4. Compute matrix-matrix multiplication between the transpose of X and X (XTX)
XTX = np.dot(X.T, X)

# 5. Compute the inverse of XTX
XTX_inv = np.linalg.inv(XTX)

# 6. Create an array y with values [1100, 1300, 800, 900, 1000, 1100]
y = np.array([1100, 1300, 800, 900, 1000, 1100])

# 7. Multiply the inverse of XTX with the transpose of X, and then multiply by y to get w
w = np.dot(np.dot(XTX_inv, X.T), y)

# 8. Compute the sum of all elements of the result (w)
sum_w = np.sum(w)

print("Sum of all elements in w:", sum_w)


Sum of all elements in w: 91.2998806299555
